docs/rfcs/2024-07-04-conversational-agent.md

```diff
@@ -0,0 +1,55 @@
+
+Historically, we tried different UX approaches with various amount of explanations texts, sheets that open when you press (i) icon, and external documentations.
+
+All these approaches are ineffective, and we lose the new users during the onboarding, especially if they don't come via the introduction of the existing users - they simply fail to understand why the app design is what it is, why it cannot do what every other app is doing, and instead of using the onboarding as a promotion opportunity, we simply lose the users.
+
+At this point, as the users are growing it appears to be the #1 blocker for the growth.
+
+## Solution
+
+We believe that the best approach to both onboard the users and to explain all the complex concepts while using the app is conversational assistant that offers scripted conversation with both multiple choice and free-form questions that users can interact with and both understand what options and choices best fit their needs and manage the app configuration and actions via conversational UX.
+
+## Implementation
+
+Possible options are:
+
+## Just write code in the UI
+
+Pros: it avoids any abstractions that can prove to be inadequate to the task as the requirements are discovered and evolve.
+
+Cons: even though there are some differences between the platform, particularly around how the user connects to the network, there is a lot of overlap, and this approach would result in a lot of repetive code.
+
+## Use JavaScript as a cross-platform language
+
+Pros: the same code will run on all platforms.
+
+Cons: requires complex integration and design of JS host environment to integrate both with the code code and with the UI.
+
+While this option is very tempting, and JS engine can be used for other things - chat widgets and interactive activities - it is probably unnecessary for this problem.
+
+## Programm it in Haskell
+
+Pros: no new languages, easy to create high level DSL to describe conversational scenarios, and at the same time have direct access to conversation data in Haskell code, without the need for any additional API, as would be required with JavaScript.
+
+Cons:
+- UIs would have to support protocol for UI actions and settings.
+- Harder to use to some developers in the team.
+
+Overall, this appears to be the simplest and the most effective approach.
+
+Conversation(s) with the agent would be stored in the usual database, and the agent would be interactive - it will be called by the code when the message is added to this conversation.
+
+Questions to consider:
+- ux:
+  - should it be one conversation for all user profiles or separate conversations per profile? The latter seems safer from privacy and security points of view.
+- types:
+  - conversation type. Could use "direct" or may be a special type.
```

Comment:
Local is more close, with direct there would be lots of parallel ifs and/or uncaught bugs. For example, if at some point it requires file functionality files would be uploaded/downloaded instead of simply being stored, etc. Maybe local as well would have unwanted interactions. For example, I don't think we'd want to allow forwarding to/from it. If it's fully scripted "conversation" with no free form I'm not even sure it requires full capabilities of conversation - like chat items, except some rendering logic that could be decomposed; and may require some other capabilities - for example remembering/"understanding" communication context/history. In that case it might be easier to have fully separate abstractions for it

Reply:
It could still be presented as "chat" in UI if necessary, like pending connections and requests are presented as chat previews, but it doesn't necessarily have to use regular chat items, conversations apis, preferences, etc.